---
title: "MonkeyScRNA"
author: "Xiangning Xue"
date: '2023-04-01'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      eval = FALSE)
```

```{bash}
srun -n1 -t02:00:00 --pty bash
module load gcc/10.2.0 r/4.2.0
R
```

## Prepare the files 

Craete HDF5SummarizedExperiment files. 

Original code is BaDoi_code/raw_code/differential_expression_analysis/older_DEG_analyses/step1_pseudoBulk_edgeRQLFDetRate_diffGene_by_cellType3.R

```{r}
xx.dat.dir= "/bgfs/ctseng/xix66/MonkeyScRNA/BaDoi/data/tidy_data/Seurat_projects"
DATADIR = "/bgfs/ctseng/xix66/MonkeyScRNA/output"
h5Seurat.dat.fn = "OUD_Striatum_refined_all_SeuratObj_N16.h5Seurat"
# h5Seurat.dat.fn = "OUD_Striatum_refined_msn_SeuratObj_N16.h5Seurat"

library(here)
# 1) load in cell type labels for label transfer
## read in Logan BU snRNA dataset to label transfer
save_merged_fn = here(xx.dat.dir, h5Seurat.dat.fn)
## load only the scaled and raw RNA counts
obj = save_merged_fn %>% LoadH5Seurat(assay = 'RNA') 
head(obj[[]]) #12171 samples in all and 6843 in msn

## convert to SingleCellExperiment for GlmGamPoi
sce = as.SingleCellExperiment(obj)

##############################################################
# 2) compute per-sample per cell type pseudobulk DGE profiles
## save to hdf5 file format for quick save/load
h5Dir =here(DATADIR, 'HDF5Array'); dir.create(h5Dir, showWarnings = F)
saveHDF5SummarizedExperiment(sce, h5Dir, prefix=h5Seurat.dat.fn, replace=TRUE)
```



## step1

BaDoi's original code: BaDoi_code/raw_code/differential_expression_analysis/step1_pseudoBulk_voomLimma_diffGene_bigModelSVA.OUDincelltype3.R


```{r}
## conda activate r4
## packages for data table processing 
library(here)
library(tidyverse)
library(rlang)
library(writexl)

## main Seurat package snRNA-seq pacakges
library(Seurat)
library(SeuratDisk)
library(future)

## main differential gene expression package
library(SingleCellExperiment)
library(DelayedArray)
library(HDF5Array)
library(Matrix.utils)
library(limma)
library(edgeR)
library(sva)
# BiocManager::install("swfdr")
library(swfdr)

## regress out the surrogate variables
# BiocManager::install("LieberInstitute/jaffelab")
library(jaffelab)

ss <- function(x, pattern, slot = 1, ...) { 
  sapply(strsplit(x = x, split = pattern, ...), '[', slot) }
options(stringsAsFactors = F)
options(repr.plot.width=11, repr.plot.height=8.5)

# xx.dat.dir= "/bgfs/ctseng/xix66/MonkeyScRNA/BaDoi/data/tidy_data/Seurat_projects"
DATADIR = "/bgfs/ctseng/xix66/MonkeyScRNA/output"
h5Seurat.dat.fn = "OUD_Striatum_refined_all_SeuratObj_N16.h5Seurat"
pseudo.fn = "OUD_Striatum_refined_all_PseudoBulk_N16.sce2.rds"
sva.fn = "OUD_Striatum_refined_all_PseudoBulk_N16.sva.rds"
voom.fn = "voomLimma_norm_object_N16pb.rds"
voom.sva.fn = "voomLimma_norm_sva_object_N16pb.rds"
lmfit.fn = "voomLimma_diffGene_bigModelFit.rds"
lmfit.sva.fn = "voomLimma_diffGene_bigModelFitSVA.rds"
design.fn = "bigModelFit_designMatrix.rds"
design.sva.fn = "bigModelFitSVA_designMatrix.rds"
celltype_prop.fn = "rdas/OUD_Striatum_refined_celltype3_proportion.rds"
sumtab.fn = "Figures/voomLimma_nDE_pdf"
sumtab.sva.fn = "Figures/voomLimma_sva_nDE_pdf"
DElist.fn = paste0("Monkey_voom", c("_", "_sva_"), "limma_bigModelSVA_N16.celltype.rds")
DEsheet.fn = paste0("Monkey_voom", c("_", "_sva_"), "limma_bigModelSVA_N16.celltype.xlsx")
DEup.fn = paste0("Monkey_voom", c("_", "_sva_"), "limma_bigModelSVA_N16.celltype.lowConfCutOff.upReg.xlsx")
DEdown.fn = paste0("Monkey_voom", c("_", "_sva_"), "limma_bigModelSVA_N16.celltype.lowConfCutOff.dnReg.xlsx")

#######################################################
# 0) Seurat uses the future package for parallelization
## set to be parallel over 28 cores
plan("multicore", workers = 28)
options(future.globals.maxSize = 80 * 1024^3)


##################################################
# 1) create or load pseudobulk sce object
save_pseudobulk =here(DATADIR, 'rdas', pseudo.fn)
if(!file.exists(save_pseudobulk)){
  ## load the single cell counts
  h5Dir =here(DATADIR, 'HDF5Array'); dir.create(h5Dir, showWarnings = F)
  sce = loadHDF5SummarizedExperiment(h5Dir, prefix=h5Seurat.dat.fn)
  
  ## merge interneurons again
  sce$celltype3 = ifelse(grepl('Int', sce$celltype3), 'Interneuron',sce$celltype3)
  #calculate cell proportion 
  a.tab = table(sce$celltype3, sce$Monkey)
  cell_prop = a.tab %>% apply(2, function(x){x/sum(x)}) %>% apply(1, mean)
  #(sce$celltype3 %>% table )/length(sce$celltype3 ) actually very close
  sum(cell_prop)==1
  saveRDS(cell_prop, paste0(DATADIR, "/", celltype_prop.fn))
  
  ## correct the sex according to RM_metadata_030923
  sce$Sex = ifelse(sce$Monkey %in% c("S1", "S3"), "Female", "Male")
  
  ## aggregate by cluster-sample to create pseudo-bulk count matrix
  colData(sce)
  # groups <- colData(sce)[, c("celltype3", "Case", 'Region')]
  groups <- colData(sce)[, c("celltype3", "Monkey")]
  pb <- aggregate.Matrix(t(counts(sce)), groupings = groups, fun = "sum") 
  dim(pb) #72 22138
  
  ## split by cluster, transform & rename columns
  pb_colData = colData(sce) %>% as.data.frame() %>%
    rownames_to_column('match') %>% 
    mutate(Pair = factor(orig.ident), match = paste(celltype3, Monkey, sep = '_')) %>%  #what is pair?
    # mutate(Pair = factor(Pair), match = paste(celltype3, Case, Region, sep = '_')) %>% 
    filter(!duplicated(match)) %>% column_to_rownames('match')
  pb_colData = pb_colData[rownames(pb),]
  
  ## make sure this PD is correct
  # with(pb_colData, table(Case, celltype3, Region))
  with(pb_colData, table(Monkey, celltype3))
  
  ## add number of cells per aggregate
  num_cells = groups %>% as.data.frame() %>% 
    # mutate(tmp = paste(celltype3, Case, Region,sep= '_')) %>% 
      mutate(tmp = paste(celltype3, Monkey,sep= '_')) %>% 
    pull(tmp) %>% table()
  num_cells %>% as.numeric() %>% summary() #should we filter some out?
  pb_colData$numCells = num_cells[rownames(pb_colData)]
  
  ## add the gene detection rate
  pb_colData$cdr <- scale(rowMeans(pb > 0)) 
  
  ## create SingleCellExperiment from pseudo bulk counts across all cell types and region
  (pb <- SingleCellExperiment(assays = t(pb), colData = pb_colData))
  
  ## remap case index nested inside OUD dx
  # remap_case_idx = split(pb$Case, pb$DSM.IV.OUD) %>% 
  remap_case_idx = split(pb$Monkey, pb$condition) %>% 
    lapply(function(x){
      x = setNames(LETTERS[as.numeric(factor(x))], x)
      x[!duplicated(x)]
    }) %>% unlist()
  names(remap_case_idx) = ss(names(remap_case_idx), '\\.', 2)
  # pb$CaseIdx = remap_case_idx[as.character(pb$Case)]
  pb$CaseIdx = remap_case_idx[as.character(pb$Monkey)]
  
  ## check this is correct
  # table(pb$DSM.IV.OUD, pb$CaseIdx)
  # table(pb$celltype3, pb$CaseIdx)
  # table(pb$celltype3, pb$Case)
  table(pb$condition, pb$CaseIdx)
  table(pb$celltype3, pb$CaseIdx)
  table(pb$celltype3, pb$Monkey)
  
  saveRDS(pb, save_pseudobulk)
} else {
  pb = readRDS(save_pseudobulk)
}

####################################################
## 2) filter pseudobulk samples that have too few cells
pb = pb[, pb$numCells > 15]# 70 out of 72 
# pb = pb[, pb$celltype3 != 'Mural'] # drop mural cells b/c too few
pb$celltype3 = make.names(pb$celltype3) %>% as.factor()
# pb$Region = as.factor(pb$Region)
pb$Sex = as.factor(pb$Sex)
pb$numCells = as.numeric(pb$numCells)

## make an interaction term for all the combinations
# pb$celltype_dx_rg_sex = interaction(pb$DSM.IV.OUD, pb$celltype3,pb$Sex, pb$Region) %>% 
pb$celltype_dx = interaction(pb$condition, pb$celltype3) %>% #female samples are too few to consider sex effect
  as.factor() %>% droplevels()
table(pb$celltype_dx)

## interaction term w/o the DX
# pb$celltype_rg_sex = interaction(pb$celltype3,pb$Sex, pb$Region)  %>% 
pb$celltype0 = interaction(pb$celltype3)  %>% 
  as.factor() %>% droplevels()
table(pb$celltype0)

## construct design & contrast matrix regressing out the DetRate
design <- model.matrix(~ 0 + celltype_dx  + # term capturing the main effects
                         # Age + PMI + RIN + numCells + cdr, # co-variates 
                         numCells, # co-variates 
                       data = colData(pb))

## construct the null model, used in regressing out the various factors
design0 <- model.matrix(~ 0 +  celltype0  + # term capturing the effects w/o Dx
                          # Age + PMI + RIN  + numCells + cdr, # co-variates 
                          numCells, # co-variates 
                        data = colData(pb))

####################################
# 3) normalization using voom-limma
y <- DGEList(counts = assays(pb)[[1]])
dim(y) # 22138    70

## filter out genes w/ low counts
A <- rowMeans(y$counts)
isexpr <- A > 5
y <- y[isexpr, , keep.lib.size = FALSE]
dim(y) # 7064   210

## filter out ribosomal genes, filter out mitochondria genes
drop.genes <- grep("^RP[SL]|^MT-",rownames(y), value = T, invert = F)
drop.genes %>% sort %>% data.frame() %>% 
  write_tsv(here(DATADIR, 'tables', 'dropped_mito_ribo_genes.tsv'))
keep.genes <- grep("^RP[SL]|^MT-",rownames(y), value = T, invert = T)
y = y[keep.genes, , keep.lib.size = FALSE]
dim(y) #  6980   210

# normalize counts
y <- calcNormFactors(y)

## voom precision weights and sample-wise quality weights normalization
v <- voomWithQualityWeights(y, design)
# cor <- duplicateCorrelation(v, design, block = colData(pb)$Case)
cor <- duplicateCorrelation(v, design, block = colData(pb)$Monkey)
cor$consensus # 0.07450211

## recalculate weights after adjusting for correlated samples from same subject
# v <- voomWithQualityWeights(y, design, block = colData(pb)$Case, 
v <- voomWithQualityWeights(y, design, block = colData(pb)$Monkey, 
          correlation = cor$consensus)
# cor <- duplicateCorrelation(v, design, block = colData(pb)$Case)
cor <- duplicateCorrelation(v, design, block = colData(pb)$Monkey)
cor$consensus # 0.07461851

#XX: also save v without sva components
save_voom = here(DATADIR, 'rdas', voom.fn)
saveRDS(v, file = save_voom)
# v = readRDS(here::here(DATADIR, 'rdas', voom.fn))

############################################################
# 4) Use surrogate variables to estimate unmodeled variation

## estimate the number of SVs from the adjusted
# save_sva =here(DATADIR, 'rdas', 'BU_OUD_Striatum_refined_all_PseudoBulk_N22.sva.rds')
save_sva =here(DATADIR, 'rdas', sva.fn)
if(! file.exists(save_sva)){
  (n.sv = num.sv(v$E, design, method="be", seed = set.seed(1))) #23
  svobj = sva(v$E, design, design0, n.sv=n.sv, B = 20)
  saveRDS(svobj, save_sva)
} else {
  svobj = readRDS(save_sva)
}

## add the SVs to the model matrix
designSV = cbind(design, svobj$sv)
design0SV = cbind(design0, svobj$sv)

## recalculate sample quality weights after calculating the SVs
# v <- voomWithQualityWeights(y, designSV, block = colData(pb)$Case, 
v.sva <- voomWithQualityWeights(y, designSV, block = colData(pb)$Monkey, 
                            correlation = cor$consensus)
# cor <- duplicateCorrelation(v, designSV, block = colData(pb)$Case)
cor.sva <- duplicateCorrelation(v.sva, designSV, block = colData(pb)$Monkey)
cor.sva$consensus # 0.0513175

# save_voom = here(DATADIR, 'rdas', 'voomLimma_norm_object_N222pb.rds')
save_voom_sva = here(DATADIR, 'rdas', voom.sva.fn)
# saveRDS(v, file = save_voom)
saveRDS(v.sva, file = save_voom_sva)

####################################################################
input.list = list(v = list(v, v.sva), 
                  cor = list(cor, cor.sva),
                  design = list(design, designSV))
fn.list = list(lmfit.fn = list(lmfit.fn, lmfit.sva.fn), 
               design.fn = list(design.fn, design.sva.fn), 
               sumtab.fn = list(sumtab.fn, sumtab.sva.fn), 
               DElist.fn = DElist.fn, 
               DEsheet.fn = DEsheet.fn, 
               DEup.fn = DEup.fn, 
               DEdown.fn = DEdown.fn)
p.all = list()
for(i in 1:2){

  ##XX: 6.1 fit the model without SVA components
  ## 6.2 fit the model with SVA components
  fit <- lmFit(input.list$v[[i]], input.list$design[[i]], 
               block = colData(pb)$Monkey, correlation = input.list$cor[[i]]$consensus)
  fit <- eBayes(fit, robust = TRUE)
  
  save_fit = here(DATADIR, 'rdas', fn.list$lmfit.fn[[i]])
  saveRDS(fit, file = save_fit)

  save_design = here(DATADIR, 'rdas', fn.list$design.fn[[i]])
  saveRDS(input.list$design[[i]], file = save_design)
  # design = readRDS(here::here(DATADIR, 'rdas', fn.list$design.fn[[1]]))
  # designSV = readRDS(here::here(DATADIR, 'rdas', fn.list$design.fn[[2]]))

  ###########################################################
  ## 7) compute the differences b/t Dx within each cell type
  celltypes=levels(factor(pb$celltype3 %>% make.names()))
  designSV = input.list$design[[i]]
  designSV2 =designSV
  colnames(designSV2) = make.names(colnames(designSV2))

  ## make the cell type contrasts
  con_celltypes = sapply(setNames(celltypes, celltypes),function(cell) {
  # cell = colnames(designSV2) %>% make.names() %>% str_subset(paste0('\\.', cell, '\\.'))
  cell = colnames(designSV2) %>% make.names() %>% str_subset(paste0('\\.', cell))
  # OUD = cell %>% str_subset('OUD'); CTL = cell %>% str_subset('CTL')
  OUD = cell %>% str_subset('morphine'); CTL = cell %>% str_subset('control')

  N_OUD = OUD %>% length(); OUD = OUD %>% paste(collapse = ' + ')
  N_CTL = CTL %>% length(); CTL = CTL %>% paste(collapse = ' + ')
  paste('(',OUD,')/',N_OUD, '-(',CTL,')/',N_CTL)
  })


  ## proportion of each cell type
  # df_prop = 'data/tidy_data/tables/BU_OUD_Striatum_refined_celltype3_proportions.txt' %>% 
  #   read_tsv() %>% deframe()
  df_prop = readRDS(paste0(DATADIR, "/", celltype_prop.fn))
  names(df_prop) = names(df_prop) %>% make.names()
  df_prop = df_prop[celltypes]

  ind_neur = grepl('^D|^Int',celltypes)
  ind_glia = !grepl('^D|^Int',celltypes)

  ## create the contrasts for OUD effect Between all cells or major classes
  con_groups = c('All' = paste0('(', con_celltypes,')*', df_prop) %>% paste(collapse = ' + '), 
                 'Neuron' = paste0('(', con_celltypes[ind_neur],')*', 
                                   df_prop[ind_neur]/sum(df_prop[ind_neur])) %>% paste(collapse = ' + '), 
                 'Glia' =  paste0('(', con_celltypes[ind_glia],')*', 
                                df_prop[ind_glia]/sum(df_prop[ind_glia])) %>% paste(collapse = ' + '))

  ## refit the model based on these contrasts
  cont.matrix <- makeContrasts(contrasts= c(con_groups, con_celltypes), levels=designSV2)
  rownames(cont.matrix) = colnames(designSV)
  fit2 <- contrasts.fit(fit, cont.matrix) %>% eBayes()

  ## compute the DEGs from these contrasts
  deg_list = lapply(setNames(colnames(cont.matrix),  names(c(con_groups, con_celltypes))), 
                    function(coef){
  topTable(coef = coef, fit =fit2, n=Inf) %>% arrange(P.Value) %>% 

  ## use SWFDR to increase power of detecting DEGs based on avg expression covariate
  ## https://pubmed.ncbi.nlm.nih.gov/30581661/
  mutate(adj.P.Val.Within =  lm_qvalue(P.Value, X=AveExpr)$q) %>%
  dplyr::select(-adj.P.Val) %>% rownames_to_column('gene') 
                      })

  ## FDR correction Between all tests
  deg_list = deg_list %>% data.table::rbindlist(idcol = 'celltype') %>% 
    mutate(adj.P.Val.Between =  lm_qvalue(P.Value, X=AveExpr)$q) %>%
    split(by = 'celltype')

  # FDR cutoff
  sapply(deg_list, function(x) x[x$adj.P.Val.Within < 0.05,] %>% nrow())
  sapply(deg_list, function(x) x[x$adj.P.Val.Between < 0.05,] %>% nrow())

  # lower confidence cutoff
  sapply(deg_list, function(x) x[x$P.Value < 0.01,] %>% nrow())

  sum.tab = rbind.data.frame(sapply(deg_list, function(x) x[x$adj.P.Val.Within < 0.05,] %>% nrow()), 
                             sapply(deg_list, function(x) x[x$adj.P.Val.Between < 0.05,] %>% nrow()), 
                             sapply(deg_list, function(x) x[x$P.Value < 0.01,] %>% nrow()))
  colnames(sum.tab) = names(deg_list)
  rownames(sum.tab) = c("adj.P.Val.Within < 0.05", "adj.P.Val.Between < 0.05", "P.Value < 0.01")

  pdf(paste0(DATADIR, "/", fn.list$sumtab.fn[[i]]), height=3, width=15)
  gridExtra::grid.table(sum.tab)
  dev.off()

  ####################################################################
  ## 8) save the output of voom_limma differential state analyses
  rdasDir =file.path(DATADIR, 'rdas'); dir.create(rdasDir, showWarnings = F)
  save_res_fn = here(rdasDir, fn.list$DElist.fn[i])
  saveRDS(deg_list, save_res_fn)

  tablesDir =file.path(DATADIR, 'tables'); dir.create(tablesDir, showWarnings = F)
  save_res_fn2 = here(tablesDir, fn.list$DEsheet.fn[i])
  deg_list %>% lapply(function(x) x %>% arrange(P.Value)) %>% writexl::write_xlsx(save_res_fn2)

  # save tables DEGs w/ P.Value < alpha up and down regulated
  save_res_fn3 = here(tablesDir, fn.list$DEup.fn[i])
  deg_list %>% lapply(function(x) x %>% arrange(P.Value) %>% filter(P.Value < 0.01, logFC > 0)) %>% 
  writexl::write_xlsx(save_res_fn3)

  save_res_fn4 = here(tablesDir, fn.list$DEdown.fn[i])
  deg_list %>% lapply(function(x) x %>% arrange(P.Value) %>% filter(P.Value < 0.01, logFC < 0)) %>% 
  writexl::write_xlsx(save_res_fn4)
  
  p.all[[i]] = do.call(rbind.data.frame, lapply(deg_list, function(a.list){
    data.frame(celltype = a.list[, 1], gene = a.list[, 2], P.value = a.list[, "P.Value"])
  }))
}

saveRDS(p.all, paste0(DATADIR, "/rdas/", "pvals_all.rds"))
p.all = readRDS(paste0(DATADIR, "/rdas/", "pvals_all.rds"))
p.all[[1]]$model = "limma"
p.all[[2]]$model = "limma_sva"
p.all = rbind.data.frame(p.all[[1]], p.all[[2]])
library(ggplot2)
#make histogram
pdf(paste0(DATADIR, "/Figures/", "pval_dist.pdf"),height = 12*4, width = 2*4)
ggplot(p.all, aes(x = P.Value))+
  geom_histogram()+
  facet_grid(celltype~model)
dev.off()

#make pair-wise scatter plots
p.all = readRDS(paste0(DATADIR, "/rdas/", "pvals_all.rds"))
colnames(p.all[[1]])[3] = "p_limma"
colnames(p.all[[2]])[3] = "p_limma_sva"
p.all2 = inner_join(p.all[[1]], p.all[[2]])
library(ggplot2)
pdf(paste0(DATADIR, "/Figures/", "pval_scatter.pdf"),height = 3*4, width = 4*4)
ggplot(p.all2, aes(x = -log10(p_limma), y = -log10(p_limma_sva)))+
  geom_point(alpha = 0.5)+
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "gold")+
  facet_wrap(~celltype)
dev.off()

pdf(paste0(DATADIR, "/Figures/", "pval_scatter2.pdf"),height = 3*4, width = 4*4)
ggplot(p.all2, aes(x = -log10(p_limma), y = -log10(p_limma_sva)+log10(p_limma)))+
  geom_point(alpha = 0.5)+
  geom_abline(intercept = 0, slope = 0, linetype = "dashed", color = "gold")+
  facet_wrap(~celltype)
dev.off()

# mean((p.all2 %>% mutate(x = p_limma_sva-p_limma) %>% filter(celltype=="All") %>% pull(x))>0)

#######################################
## 9) check out some interesting genes
rdasDir =file.path(DATADIR, 'rdas')
save_res_fn = here::here(rdasDir, DElist.fn[2])
deg_list = readRDS(save_res_fn)
sapply(deg_list, function(x){
  x %>% filter(gene %in% c('CRHR1', 'LRP8')) %>% pull(adj.P.Val.Between)
})

lapply(deg_list, function(x){
  x %>% filter(adj.P.Val.Between < 0.05) %>% pull(gene) %>% paste(collapse = ', ')
})

lapply(deg_list, function(x){
  x %>% filter(adj.P.Val.Between < 0.05) %>% pull(gene) %>% 
    str_subset('^FOX|^GAD|^CHAT$|^TH$|^KCN|^SCN|^SLC') %>% 
    paste(collapse = ', ')
})

lapply(deg_list, function(x){
  x %>% filter(adj.P.Val.Between < 0.05) %>% pull(gene) %>% 
    str_subset('^DRD|^OPR|^CADM|^CRHR') %>% paste(collapse = ', ')
})
```

## step6

Original code: BaDoi_code/raw_code/differential_expression_analysis/step6_gsea_pathway_enrichments.R

```{r}
## conda activate r4
## packages for data table processing 
library(here)
library(tidyverse)
library(RColorBrewer)
# install.packages("rcartocolor")
library(rcartocolor)
library(ggpubr)

library(data.table)
library(fgsea)
# BiocManager::install("swfdr")
library(swfdr)
# install.packages("msigdbr")
library(msigdbr)

ss <- function(x, pattern, slot = 1, ...) { 
  sapply(strsplit(x = x, split = pattern, ...), '[', slot) }
options(stringsAsFactors = F)
options(repr.plot.width=11, repr.plot.height=8.5)

DATADIR = "output"
DElist.fn = paste0(DATADIR,"/rdas/Monkey_voom", "_sva_", "limma_bigModelSVA_N16.celltype.rds")

## make for this subdirs
# PLOTDIR = "/bgfs/ctseng/xix66/MonkeyScRNA/output/Figures/exploratory"
PLOTDIR = "output/Figures/exploratory"
paste0(PLOTDIR, "/", c('plots', 'tables', 'rdas')) %>% sapply(dir.create, showWarnings = F, recursive = TRUE)

############################################################
# 1) read in the DEG lists per comparison of OUD vs. Control
res.celltype = DElist.fn %>% readRDS()
names(res.celltype) = paste0(names(res.celltype), '#All')

deg_rank_list1 = lapply(res.celltype, function(deg){
  deg %>% mutate(tmp = -log10(P.Value) * sign(logFC)) %>% 
    dplyr::select(gene, tmp) %>% arrange(tmp) %>% deframe()
})

deg_rank_list = deg_rank_list1

#############################################
## 2) get gene ontologies, use human genes
## grab the H, Hallmark set of gene pathways
## grab the C2, which is the curated canonical pathway sets
## grab the C5, which is the Gene Ontology sets
## https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp
pathways_df =  bind_rows(msigdbr("human", category="H"), 
                         msigdbr("human", category="C2"), 
                         msigdbr("human", category="C5"))
"Macaca mulatta"
## get the SynGO gene ontologies, use human genes
syngo_df = readxl::read_xlsx("data/SynGO_bulk_download_release_20210225/syngo_annotations.xlsx")
pathways_df = rbindlist(list(pathways_df, syngo_df), fill = T) 

## reshape/label for pathway naming purposes
pathways <-pathways_df %>% 
  mutate(
    gs_subcat = ifelse(is.na(gs_subcat) | gs_subcat == '', gs_cat, gs_subcat),
    gs_name = paste(gs_subcat, gs_name, sep ='#')) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

## exclude the really really big gene sets
lengths(pathways) %>% summary()
pathways = pathways[lengths(pathways)<500]
length(pathways) # 21058

table(pathways_df$gs_cat)

pathways_df2 = pathways_df %>% 
  dplyr::select(gs_subcat, gs_name, gs_description) %>% distinct() %>% 
  dplyr::rename('pathway_group' = 'gs_subcat', 'pathway' = 'gs_name',
                'description' = 'gs_description')

## conduct the GSEA analyses
gsea_list = lapply(deg_rank_list, fgsea, pathways = pathways,
                   minSize=15, ## minimum gene set size
                   maxSize=400) ## maximum gene set size

alpha = 0.05
gsea_df = gsea_list %>% rbindlist(idcol = 'group') %>% 
  arrange(pval) %>% filter(!is.na(pval)) %>% 
  mutate(
    MSigDb_Group = ss(pathway, '#', 1), 
    celltype = group %>% ss('#', 1), 
    MORPHINE.v.CTL.in = group %>% ss('#', 1),
    pathway = ss(pathway, '#', 2),
    padj = lm_qvalue(pval, X=size)$q, 
    celltype = group %>% ss('#', 1), 
    leadingEdge = map_chr(leadingEdge, paste, collapse = ',')) %>% 
  inner_join(pathways_df2) %>% dplyr::select(-group) %>% 
  relocate(MORPHINE.v.CTL.in, celltype, MSigDb_Group, description, .before= everything()) %>% 
  split(f = .$MORPHINE.v.CTL.in)

## save the enrichment w/ all the pathways, significant and otherwise
gsea_df %>% saveRDS(paste0("output/rdas/Monkey_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.unfiltered.rds"))
# gsea_df = readRDS(paste0("output/rdas/Monkey_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.unfiltered.rds"))
## filter out just the significant pathways
gsea_df = lapply(gsea_df, function(x){
  x %>% filter(padj < alpha)
})
sapply(gsea_df, nrow)

out_fn = paste0('output', 
              '/tables/Monkey_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.xlsx')
gsea_df %>% writexl::write_xlsx(out_fn)
gsea_df %>%saveRDS(paste0("output/rdas/Monkey_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.rds"))

## take a look at the enrichments
gsea_df2 = gsea_df %>% rbindlist()
a.tab = table(gsea_df2$MORPHINE.v.CTL.in, gsea_df2$MSigDb_Group)

pdf(paste0("output/Figures/Monkey_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.pdf"), height=6, width=10)
gridExtra::grid.table(a.tab, rows = rownames(a.tab), cols = colnames(a.tab))
dev.off()

#do the same for monkey pathways
msigdbr_species()
pathways_df =  bind_rows(msigdbr("Macaca mulatta", category="H"), 
                         msigdbr("Macaca mulatta", category="C2"), 
                         msigdbr("Macaca mulatta", category="C5"))

## reshape/label for pathway naming purposes
pathways <-pathways_df %>% 
  mutate(
    gs_subcat = ifelse(is.na(gs_subcat) | gs_subcat == '', gs_cat, gs_subcat),
    gs_name = paste(gs_subcat, gs_name, sep ='#')) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

## exclude the really really big gene sets
lengths(pathways) %>% summary()
pathways = pathways[lengths(pathways)<500]
length(pathways) # 21240

table(pathways_df$gs_cat)

pathways_df2 = pathways_df %>% 
  dplyr::select(gs_subcat, gs_name, gs_description) %>% distinct() %>% 
  dplyr::rename('pathway_group' = 'gs_subcat', 'pathway' = 'gs_name',
                'description' = 'gs_description')

## conduct the GSEA analyses
gsea_list = lapply(deg_rank_list, fgsea, pathways = pathways,
                   minSize=15, ## minimum gene set size
                   maxSize=400) ## maximum gene set size

alpha = 0.05
gsea_df = gsea_list %>% rbindlist(idcol = 'group') %>% 
  arrange(pval) %>% filter(!is.na(pval)) %>% 
  mutate(
    MSigDb_Group = ss(pathway, '#', 1), 
    celltype = group %>% ss('#', 1), 
    MORPHINE.v.CTL.in = group %>% ss('#', 1),
    pathway = ss(pathway, '#', 2),
    padj = lm_qvalue(pval, X=size)$q, 
    celltype = group %>% ss('#', 1), 
    leadingEdge = map_chr(leadingEdge, paste, collapse = ',')) %>% 
  inner_join(pathways_df2) %>% dplyr::select(-group) %>% 
  relocate(MORPHINE.v.CTL.in, celltype, MSigDb_Group, description, .before= everything()) %>% 
  split(f = .$MORPHINE.v.CTL.in)

## save the enrichment w/ all the pathways, significant and otherwise
gsea_df %>% saveRDS(paste0("output/rdas/Monkey_GSEA_enrichment_in_monkey_msigdb_H_C2_C5.unfiltered.rds"))
## filter out just the significant pathways
gsea_df = lapply(gsea_df, function(x){
  x %>% filter(padj < alpha)
})
sapply(gsea_df, nrow)

out_fn = paste0('output', 
              '/tables/Monkey_GSEA_enrichment_in_monkey_msigdb_H_C2_C5.xlsx')
gsea_df %>% writexl::write_xlsx(out_fn)
gsea_df %>%saveRDS(paste0("output/rdas/Monkey_GSEA_enrichment_in_monkey_msigdb_H_C2_C5.rds"))

## take a look at the enrichments
gsea_df2 = gsea_df %>% rbindlist()
a.tab = table(gsea_df2$MORPHINE.v.CTL.in, gsea_df2$MSigDb_Group)

pdf(paste0("output/Figures/Monkey_GSEA_enrichment_in_monkey_msigdb_H_C2_C5_SynGO.pdf"), height=6, width=10)
gridExtra::grid.table(a.tab, rows = rownames(a.tab), cols = colnames(a.tab))
dev.off()
```

### repeat without sva
```{r}
## conda activate r4
## packages for data table processing 
library(here)
library(tidyverse)
library(RColorBrewer)
# install.packages("rcartocolor")
library(rcartocolor)
library(ggpubr)

library(data.table)
library(fgsea)
# BiocManager::install("swfdr")
library(swfdr)
# install.packages("msigdbr")
library(msigdbr)

ss <- function(x, pattern, slot = 1, ...) { 
  sapply(strsplit(x = x, split = pattern, ...), '[', slot) }
options(stringsAsFactors = F)
options(repr.plot.width=11, repr.plot.height=8.5)

DATADIR = "output"
DElist.fn = paste0(DATADIR,"/rdas/Monkey_voom", "_", "limma_bigModelSVA_N16.celltype.rds")

## make for this subdirs
# PLOTDIR = "/bgfs/ctseng/xix66/MonkeyScRNA/output/Figures/exploratory"
PLOTDIR = "output/Figures/exploratory"
paste0(PLOTDIR, "/", c('plots', 'tables', 'rdas')) %>% sapply(dir.create, showWarnings = F, recursive = TRUE)

############################################################
# 1) read in the DEG lists per comparison of OUD vs. Control
res.celltype = DElist.fn %>% readRDS()
names(res.celltype) = paste0(names(res.celltype), '#All')

deg_rank_list1 = lapply(res.celltype, function(deg){
  deg %>% mutate(tmp = -log10(P.Value) * sign(logFC)) %>% 
    dplyr::select(gene, tmp) %>% arrange(tmp) %>% deframe()
})

deg_rank_list = deg_rank_list1

#############################################
## 2) get gene ontologies, use human genes
## grab the H, Hallmark set of gene pathways
## grab the C2, which is the curated canonical pathway sets
## grab the C5, which is the Gene Ontology sets
## https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp
pathways_df =  bind_rows(msigdbr("human", category="H"), 
                         msigdbr("human", category="C2"), 
                         msigdbr("human", category="C5"))
"Macaca mulatta"
## get the SynGO gene ontologies, use human genes
syngo_df = readxl::read_xlsx("data/SynGO_bulk_download_release_20210225/syngo_annotations.xlsx")
pathways_df = rbindlist(list(pathways_df, syngo_df), fill = T) 

## reshape/label for pathway naming purposes
pathways <-pathways_df %>% 
  mutate(
    gs_subcat = ifelse(is.na(gs_subcat) | gs_subcat == '', gs_cat, gs_subcat),
    gs_name = paste(gs_subcat, gs_name, sep ='#')) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

## exclude the really really big gene sets
lengths(pathways) %>% summary()
pathways = pathways[lengths(pathways)<500]
length(pathways) # 21058

table(pathways_df$gs_cat)

pathways_df2 = pathways_df %>% 
  dplyr::select(gs_subcat, gs_name, gs_description) %>% distinct() %>% 
  dplyr::rename('pathway_group' = 'gs_subcat', 'pathway' = 'gs_name',
                'description' = 'gs_description')

## conduct the GSEA analyses
gsea_list = lapply(deg_rank_list, fgsea, pathways = pathways,
                   minSize=15, ## minimum gene set size
                   maxSize=400) ## maximum gene set size

alpha = 0.05
gsea_df = gsea_list %>% rbindlist(idcol = 'group') %>% 
  arrange(pval) %>% filter(!is.na(pval)) %>% 
  mutate(
    MSigDb_Group = ss(pathway, '#', 1), 
    celltype = group %>% ss('#', 1), 
    MORPHINE.v.CTL.in = group %>% ss('#', 1),
    pathway = ss(pathway, '#', 2),
    padj = lm_qvalue(pval, X=size)$q, 
    celltype = group %>% ss('#', 1), 
    leadingEdge = map_chr(leadingEdge, paste, collapse = ',')) %>% 
  inner_join(pathways_df2) %>% dplyr::select(-group) %>% 
  relocate(MORPHINE.v.CTL.in, celltype, MSigDb_Group, description, .before= everything()) %>% 
  split(f = .$MORPHINE.v.CTL.in)

## save the enrichment w/ all the pathways, significant and otherwise
gsea_df %>% saveRDS(paste0("output/rdas/Monkey_nosva_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.unfiltered.rds"))
# gsea_df = readRDS(paste0("output/rdas/Monkey_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.unfiltered.rds"))
## filter out just the significant pathways
gsea_df = lapply(gsea_df, function(x){
  x %>% filter(padj < alpha)
})
sapply(gsea_df, nrow)

out_fn = paste0('output', 
              '/tables/Monkey_nosva_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.xlsx')
gsea_df %>% writexl::write_xlsx(out_fn)
gsea_df %>%saveRDS(paste0("output/rdas/Monkey_nosva_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.rds"))

## take a look at the enrichments
gsea_df2 = gsea_df %>% rbindlist()
a.tab = table(gsea_df2$MORPHINE.v.CTL.in, gsea_df2$MSigDb_Group)

pdf(paste0("output/Figures/Monkey_nosva_GSEA_enrichment_in_human_msigdb_H_C2_C5_SynGO.pdf"), height=6, width=10)
gridExtra::grid.table(a.tab, rows = rownames(a.tab), cols = colnames(a.tab))
dev.off()

#do the same for monkey pathways
msigdbr_species()
pathways_df =  bind_rows(msigdbr("Macaca mulatta", category="H"), 
                         msigdbr("Macaca mulatta", category="C2"), 
                         msigdbr("Macaca mulatta", category="C5"))

## reshape/label for pathway naming purposes
pathways <-pathways_df %>% 
  mutate(
    gs_subcat = ifelse(is.na(gs_subcat) | gs_subcat == '', gs_cat, gs_subcat),
    gs_name = paste(gs_subcat, gs_name, sep ='#')) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

## exclude the really really big gene sets
lengths(pathways) %>% summary()
pathways = pathways[lengths(pathways)<500]
length(pathways) # 21240

table(pathways_df$gs_cat)

pathways_df2 = pathways_df %>% 
  dplyr::select(gs_subcat, gs_name, gs_description) %>% distinct() %>% 
  dplyr::rename('pathway_group' = 'gs_subcat', 'pathway' = 'gs_name',
                'description' = 'gs_description')

## conduct the GSEA analyses
gsea_list = lapply(deg_rank_list, fgsea, pathways = pathways,
                   minSize=15, ## minimum gene set size
                   maxSize=400) ## maximum gene set size

alpha = 0.05
gsea_df = gsea_list %>% rbindlist(idcol = 'group') %>% 
  arrange(pval) %>% filter(!is.na(pval)) %>% 
  mutate(
    MSigDb_Group = ss(pathway, '#', 1), 
    celltype = group %>% ss('#', 1), 
    MORPHINE.v.CTL.in = group %>% ss('#', 1),
    pathway = ss(pathway, '#', 2),
    padj = lm_qvalue(pval, X=size)$q, 
    celltype = group %>% ss('#', 1), 
    leadingEdge = map_chr(leadingEdge, paste, collapse = ',')) %>% 
  inner_join(pathways_df2) %>% dplyr::select(-group) %>% 
  relocate(MORPHINE.v.CTL.in, celltype, MSigDb_Group, description, .before= everything()) %>% 
  split(f = .$MORPHINE.v.CTL.in)

## save the enrichment w/ all the pathways, significant and otherwise
gsea_df %>% saveRDS(paste0("output/rdas/Monkey_nosva_GSEA_enrichment_in_monkey_msigdb_H_C2_C5.unfiltered.rds"))
## filter out just the significant pathways
gsea_df = lapply(gsea_df, function(x){
  x %>% filter(padj < alpha)
})
sapply(gsea_df, nrow)

out_fn = paste0('output', 
              '/tables/Monkey_nosva_GSEA_enrichment_in_monkey_msigdb_H_C2_C5.xlsx')
gsea_df %>% writexl::write_xlsx(out_fn)
gsea_df %>%saveRDS(paste0("output/rdas/Monkey_nosva_GSEA_enrichment_in_monkey_msigdb_H_C2_C5.rds"))

## take a look at the enrichments
gsea_df2 = gsea_df %>% rbindlist()
a.tab = table(gsea_df2$MORPHINE.v.CTL.in, gsea_df2$MSigDb_Group)

pdf(paste0("output/Figures/Monkey_nosva_GSEA_enrichment_in_monkey_msigdb_H_C2_C5_SynGO.pdf"), height=6, width=10)
gridExtra::grid.table(a.tab, rows = rownames(a.tab), cols = colnames(a.tab))
dev.off()
```

